framework/src/userobjects/ProjectedStatefulMaterialNodalPatchRecovery.C

```diff
@@ -0,0 +1,223 @@
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::
+    ProjectedStatefulMaterialNodalPatchRecoveryTempl(const InputParameters & parameters)
+  : ProjectedStatefulMaterialNodalPatchRecoveryBase(parameters),
+    _qp(0),
+    _n_components(Moose::SerialAccess<T>::size()),
+    _patch_polynomial_order(
+        static_cast<unsigned int>(getParam<MooseEnum>("patch_polynomial_order"))),
+    _multi_index(MathUtils::multiIndex(_mesh.dimension(), _patch_polynomial_order)),
+    _q(_multi_index.size()),
+    _prop(getGenericMaterialProperty<T, is_ad>("property")),
+    _current_subdomain_id(_assembly.currentSubdomainID())
+{
+}
+
+template <typename T, bool is_ad>
+void
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::initialSetup()
+{
+  // get all material classes that provide properties for this object
+  _required_materials = buildRequiredMaterials();
+}
+
+template <typename T, bool is_ad>
+Real
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::nodalPatchRecovery(
+    const Point & x, const std::vector<dof_id_type> & elem_ids, std::size_t component) const
+{
+  // Before we go, check if we have enough sample points for solving the least square fitting
+  if (_q_point.size() * elem_ids.size() < _q)
+    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
+               "polynomial order or using a higher-order quadrature scheme.");
+
+  // Assemble the least squares problem over the patch
+  RealEigenMatrix A = RealEigenMatrix::Zero(_q, _q);
+  RealEigenVector b = RealEigenVector::Zero(_q);
+  for (auto elem_id : elem_ids)
+  {
+    const auto abs = libmesh_map_find(_abs, elem_id);
+    A += abs.first;
+    b += abs.second[component];
+  }
+
+  // Solve the least squares fitting
+  RealEigenVector coef = A.completeOrthogonalDecomposition().solve(b);
+
+  // Compute the fitted nodal value
+  RealEigenVector p = evaluateBasisFunctions(x);
+  return p.dot(coef);
+}
+
+template <typename T, bool is_ad>
+RealEigenVector
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::evaluateBasisFunctions(
+    const Point & q_point) const
+{
+  RealEigenVector p(_q);
+  Real polynomial;
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
+  {
+    polynomial = 1.0;
+    mooseAssert(_multi_index[r].size() == _mesh.dimension(), "Wrong multi-index size.");
+    for (unsigned int c = 0; c < _multi_index[r].size(); c++)
+      for (unsigned int p = 0; p < _multi_index[r][c]; p++)
+        polynomial *= q_point(c);
+    p(r) = polynomial;
+  }
+  return p;
+}
+
+template <typename T, bool is_ad>
+void
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::initialize()
+{
+  _abs.clear();
+}
+
+template <typename T, bool is_ad>
+void
+ProjectedStatefulMaterialNodalPatchRecoveryTempl<T, is_ad>::execute()
+{
+  if (_t_step == 0)
+    for (const auto & mat : _required_materials[_current_subdomain_id])
+      mat->initStatefulProperties(_qrule->size());
```

Comment:
As for the aux kernel, I hope this is not necessary, and if it turns out it is, a comment would be appreciated to understand why

Reply:
I don't understand the "as for the aux kernel" part. I can add a comment. calling `initStatefulProperties` manually here is necessary, otherwise we never see the "old" state at the beginning of the simulation. We want to avoid actually making the material stateful (1. it would duplicate the storage requirements, 2. it would trigger errors, where we explicitly prohibit stateful properties)